File: solutions/java/src/main/java/com/epi/ValidPalindrome.java
package com.epi;

import static java.lang.Character.isLetterOrDigit;
import static java.lang.Character.toLowerCase;

public class ValidPalindrome {
  // @include
  public static boolean isPalindrome(String s) {
    // i moves forward, and j moves backward.
    int i = 0, j = s.length() - 1;
    while (i < j) {
      // i and j both skip non-alphanumeric characters.
      while (!isLetterOrDigit(s.charAt(i)) && i < j) {
        ++i;
      }
      while (!isLetterOrDigit(s.charAt(j)) && i < j) {
        --j;
      }
      if (toLowerCase(s.charAt(i)) != toLowerCase(s.charAt(j))) {
        return false;
      }
      ++i;
      --j;
    }
    return true;
  }
  // @exclude

  public static void main(String[] args) {
    assert(isPalindrome("A man, a plan, a canal: Panama"));
    assert(!isPalindrome("race a car"));
    assert(isPalindrome("Able was I, ere I saw Elba!"));
    assert(!isPalindrome("Ray a Ray"));
    assert(isPalindrome(""));
  }
}

File: static/solutions/cpp/valid-palindrome.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <cctype>
#include <iostream>
#include <string>

using std::cout;
using std::endl;
using std::string;

// @include
bool IsPalindrome(const string& s) {
  // i moves forward, and j moves backward.
  int i = 0, j = s.size() - 1;
  while (i < j) {
    // i and j both skip non-alphanumeric characters.
    while (!isalnum(s[i]) && i < j) {
      ++i;
    }
    while (!isalnum(s[j]) && i < j) {
      --j;
    }
    if (tolower(s[i++]) != tolower(s[j--])) {
      return false;
    }
  }
  return true;
}
// @exclude

int main(int argc, char** argv) {
  assert(IsPalindrome("A man, a plan, a canal: Panama"));
  assert(!IsPalindrome("race a car"));
  assert(IsPalindrome("Able was I, ere I saw Elba!"));
  assert(!IsPalindrome("Ray a Ray"));
  assert(IsPalindrome(""));
  return 0;
}
